File: cutilslibrary/src/main/java/com/utils/library/rx/RxBus.java
//package com.utils.library.rx;
//
//import java.util.HashMap;
//
//import io.reactivex.Flowable;
//import io.reactivex.disposables.CompositeDisposable;
//import io.reactivex.disposables.Disposable;
//import io.reactivex.processors.FlowableProcessor;
//import io.reactivex.processors.PublishProcessor;
//
//public class RxBus {
//    private static volatile RxBus mInstance;
//    private FlowableProcessor<Object> mSubject;
//    private HashMap<String, CompositeDisposable> disposableHashMap;
//
//    private RxBus() {
//        //调用toSerialized()方法，保证线程安全
//        mSubject = PublishProcessor.create().toSerialized();
//    }
//
//    public static RxBus getInstance() {
//        if (mInstance == null) {
//            synchronized (RxBus.class) {
//                if (mInstance == null) {
//                    mInstance = new RxBus();
//                }
//            }
//        }
//        return mInstance;
//    }
//
//    /**
//     * 发送事件
//     *
//     * @param o
//     */
//    public void post(Object o) {
//        mSubject.onNext(o);
//    }
//
//    /**
//     * 返回指定类型的Observable实例
//     *
//     * @param type
//     * @param <T>
//     * @return
//     */
//    public <T> Flowable<T> tObservable(final Class<T> type) {
////       return mSubject.filter(new Func1<Object, Boolean>() {
////            @Override
////            public Boolean call(Object t) {
////                return type.isInstance(t);
////            }
////        }).cast(type);
//
//        return mSubject.ofType(type);
//    }
//
//    /**
//     * 是否已有观察者订阅
//     *
//     * @return
//     */
//    public boolean hasObservers() {
//        return mSubject.hasSubscribers();
//    }
//
//    /**
//     * 保存订阅后的subscription
//     *
//     * @param o
//     * @param subscription
//     */
//    public void addSubscription(Object o, Disposable subscription) {
//        if (disposableHashMap == null) {
//            disposableHashMap = new HashMap<>();
//        }
//        String key = o.getClass().getName();
//        if (disposableHashMap.get(key) != null) {
//            disposableHashMap.get(key).add(subscription);
//        } else {
//            CompositeDisposable compositeSubscription = new CompositeDisposable();
//            compositeSubscription.add(subscription);
//            disposableHashMap.put(key, compositeSubscription);
//        }
//    }
//
//    /**
//     * 取消订阅
//     *
//     * @param o
//     */
//    public void unSubscribe(Object o) {
//        if (disposableHashMap == null) {
//            return;
//        }
//
//        String key = o.getClass().getName();
//        if (!disposableHashMap.containsKey(key)) {
//            return;
//        }
//        if (disposableHashMap.get(key) != null) {
//            disposableHashMap.get(key).dispose();
//        }
//
//        disposableHashMap.remove(key);
//    }
//
//    public void removeAll() {
//        if (disposableHashMap != null && disposableHashMap.size() > 0) {
//            for (String key : disposableHashMap.keySet()) {
//                disposableHashMap.get(key).dispose();
//            }
//            disposableHashMap.clear();
//        }
//    }
//}

File: cutilslibrary/src/main/java/com/utils/library/rx/RxBus.kt
package com.utils.library.rx

import io.reactivex.rxjava3.core.Flowable
import io.reactivex.rxjava3.disposables.CompositeDisposable
import io.reactivex.rxjava3.disposables.Disposable
import io.reactivex.rxjava3.processors.FlowableProcessor
import io.reactivex.rxjava3.processors.PublishProcessor
import java.util.HashMap

//接收来自New_Update_ShopActivity
// RxBus.getInstance().addSubscription(this, RxBus.getInstance().tObservable(BaseEntity.class)
//        .observeOn(AndroidSchedulers.mainThread())
//.compose(this.bindToLifecycle())
//.subscribe(entity -> {
//    if (binding.include.swipeToLoadLayout != null)
//        binding.include.swipeToLoadLayout.setRefreshing(true);
//}
//, throwable -> refreshError(throwable.toString())
//));

//RxBus.getInstance().post(new BaseEntity());//发送到shoplistFragment更新列表

class RxBus {

    private val mSubject: FlowableProcessor<Any> = PublishProcessor.create<Any>().toSerialized()//调用toSerialized()方法，保证线程安全


    private var disposableHashMap: HashMap<String, CompositeDisposable>? = null

    /**
     * 发送事件

     * @param o
     */
    fun post(o: Any) {
        mSubject.onNext(o)
    }

    /**
     * 返回指定类型的Observable实例

     * @param type
     * *
     * @param <T>
     * *
     * @return
    </T> */
    fun <T> tObservable(type: Class<T>): Flowable<T> {
        //       return mSubject.filter(new Func1<Object, Boolean>() {
        //            @Override
        //            public Boolean call(Object t) {
        //                return type.isInstance(t);
        //            }
        //        }).cast(type);

        return mSubject.ofType(type)
    }

    /**
     * 是否已有观察者订阅

     * @return
     */
    fun hasObservers(): Boolean {
        return mSubject.hasSubscribers()
    }

    /**
     * 保存订阅后的subscription

     * @param o
     * *
     * @param subscription
     */
    fun addSubscription(o: Any, subscription: Disposable) {
        if (disposableHashMap == null) {
            disposableHashMap = HashMap<String, CompositeDisposable>()
        }
        val key = o::class.simpleName!!
        if (disposableHashMap!![key] != null) {
            disposableHashMap!![key]?.add(subscription)
        } else {
            val compositeSubscription = CompositeDisposable()
            compositeSubscription.add(subscription)
            disposableHashMap!!.put(key, compositeSubscription)
        }
    }

    /**
     * 取消订阅

     * @param o
     */
    fun unSubscribe(o: Any) {
        if (disposableHashMap == null) {
            return
        }
        val key = o.javaClass.name
        if (!disposableHashMap!!.containsKey(key)) {
            return
        }
        if (disposableHashMap!![key] != null) {
            disposableHashMap!![key]?.dispose()
        }

        disposableHashMap!!.remove(key)
    }

    fun removeAll() {
        if (disposableHashMap != null && disposableHashMap!!.size > 0) {
            for (key in disposableHashMap!!.keys) {
                disposableHashMap!![key]?.dispose()
            }
            disposableHashMap!!.clear()
        }
    }

    companion object {
        @Volatile private var mInstance: RxBus? = null

        fun getInstance(): RxBus {
            if (mInstance == null) {
                synchronized(RxBus::class.java) {
                    if (mInstance == null) {
                        mInstance = RxBus()
                    }
                }
            }
            return mInstance!!
        }
    }
}
